support/build-doc-bundle.py

```diff
@@ -0,0 +1,229 @@
+#!/usr/bin/env -S pipx run
+
+# /// script
+# requires-python = ">=3.9"
+# dependencies = ["click", "jinja2", "pyyaml"]
+# ///
+
+from __future__ import annotations
+
+import dataclasses
+import glob
+import io
+import os
+import tarfile
+import textwrap
+import typing as t
+
+try:
+    import click
+    import jinja2
+    import yaml
+except ModuleNotFoundError:
+    print(
+        """\
+library imports failed in 'build-doc-bundle.py'
+
+Did you try to run the script with 'python build-doc-bundle.py'?
+Use 'pipx run build-doc-bundle.py' or './build-doc-bundle.py' instead!
+""",
+        end="",
+    )
+    raise SystemExit(1)
+
+
+@click.option(
+    "-o",
+    "--output",
+    default="doc_bundle.tar.gz",
+    type=click.Path(writable=True, dir_okay=False, resolve_path=True),
+    help="Output filename.",
+)
+@click.command
+def main(output: str) -> None:
+    click.echo("Bundling examples for docs.globus.org", err=True)
+    go_to_repo_root()
+
+    with tarfile.open(output, "w:gz") as archive:
+        for filename, file_bytes in build_all_files():
+            file_info = tarfile.TarInfo(filename)
+            file_info.size = len(file_bytes)
+            archive.addfile(file_info, io.BytesIO(file_bytes))
+
+    click.echo(f"Done. Output is available at {output}")
+
+
+def build_all_files() -> t.Iterator[tuple[str, bytes]]:
+    all_example_configs: list[ExampleDocBuildConfig] = []
+    for config_filename in find_build_configs():
+        example_dir = os.path.dirname(config_filename)
+        click.echo(f"building for {example_dir}", err=True)
+
+        config = load_config(config_filename)
+        all_example_configs.append(config)
+
+        if not config.readme_is_index:
+            _abort("doc builds currently only support readme_is_index behavior")
+
+        for sub_path in ("definition.json", "input_schema.json", "sample_input.json"):
+            with open(os.path.join(example_dir, sub_path), "rb") as fp:
+                yield f"{config.example_dir}/{sub_path}", fp.read()
+        with open(os.path.join(example_dir, "README.adoc"), "rb") as fp:
+            content = fp.read()
+            if config.append_source_blocks:
+                content = append_source_blocks(content)
+            content = prepend_preamble(config, content)
+            yield f"{config.example_dir}/index.adoc", content
+    yield "index.adoc", build_index_doc(all_example_configs)
+
+
+def go_to_repo_root() -> None:
+    support_dir = os.path.dirname(__file__)
+    os.chdir(os.path.dirname(support_dir))
+
+
+def find_build_configs() -> t.Iterator[str]:
+    for item in glob.glob("*/**/.doc_config.yaml", recursive=True):
+        yield item
```

Comment:
Yielding strings instead of `pathlib.Path` instances has had an outsized effect on the readability and verbosity of the code. For example, this code:

```python
for sub_path in ("definition.json", "input_schema.json", "sample_input.json"):
    with open(os.path.join(example_dir, sub_path), "rb") as fp:
        yield f"{config.example_dir}/{sub_path}", fp.read()
```

could be written like this if `pathlib.Path` instances are used throughout the code:

```python
for sub_path in ("definition.json", "input_schema.json", "sample_input.json"):
    file = example_dir / sub_path
    yield file, file.read()
```

Reply:
I wouldn't pin that on this function in particular. I actually decided to stick with strings because I didn't want to mess with the conversion functionality for `click.Path`. But I'll figure that out and then we can have paths throughout. (I didn't want it _mixed_)